AzureMaps/AzureMaps.php

```diff
@@ -0,0 +1,312 @@
+    }
+
+    /**
+     * @param ReverseQuery $query
+     *
+     * @return Collection
+     *
+     * @throws \Geocoder\Exception\Exception
+     */
+    public function reverseQuery(ReverseQuery $query): Collection
+    {
+        $url = $this->buildReverseGeocodeUrl(
+            self::REVERSE_ENDPOINT_URL,
+            $query->getCoordinates()->getLatitude(),
+            $query->getCoordinates()->getLongitude()
+        );
+
+        $content = $this->getUrlContents($url);
+
+        $response = $this->validateResponse($content, $url);
+        $addresses = $this->formatReverseGeocodeResponse($response);
+
+        return new AddressCollection($addresses);
+    }
+
+    /**
+     * Returns the provider's name.
+     *
+     * @return string
+     */
+    public function getName(): string
+    {
+        return 'azure_maps';
+    }
+
+    /**
+     * Returns an array of non null geocode /reverse-geocode options
+     * @param array $options
+     */
+    private function setOptions(array $options)
+    {
+        $options = array_merge($this->options, $options);
+
+        $this->options = array_filter($options, function ($option) {
+            return !is_null($option);
+        });
+    }
+
+    /**
+     * Returns an array of keys to replace
+     *
+     * @return array
+     */
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
+    /**
+     * @param string $url
+     * @param string $query
+     * @return string
+     */
+    private function buildGeocodeUrl(string $url, string $query): string
+    {
+        $format = $this->getFormat();
+        $options = http_build_query($this->getOptions());
+        $query = urlencode($query);
+        $subscriptionKey= $this->getSubscriptionKet();
+
+        return "{$url}/$format?api-version=1.0&subscription-key={$subscriptionKey}&{$options}&query={$query}";
+    }
+
+    /**
+     * @param string $url
+     * @param string $latitude
+     * @param string $longitude
+     * @return string
+     */
+    private function buildReverseGeocodeUrl(string $url, string $latitude, string $longitude): string
+    {
+        $format = $this->getFormat();
+        $options = http_build_query($this->getOptions());
+        $subscriptionKey= $this->getSubscriptionKet();
+
+        return "{$url}/$format?api-version=1.0&subscription-key={$subscriptionKey}&{$options}&query={$latitude},{$longitude}";
+    }
+
+    /**
+     * @return string
+     */
+    private function getSubscriptionKet(): string
+    {
+        return $this->subscriptionKey;
+    }
+
+    /**
+     * @param string $subscriptionKey
+     */
+    private function setSubscriptionKey(string $subscriptionKey)
+    {
+        $this->subscriptionKey = $subscriptionKey;
+    }
+
+    /**
+     * @param string $format
+     */
+    private function setFormat(string $format)
+    {
+        $this->format = $format;
+    }
+
+    /**
+     * @return string
+     */
+    private function getFormat(): string
```

Comment:
Do we need this getter/setter as well?

Reply:
How do you decide which property needs to have a setter/getter, and which isn't?

Reply:
Unless it needs to be accessed outside of the class, no need for a getter (private getters are useless IMO)